test: simplify test_runner.py

tdb3 opened the pull request:
Implements the simplifications to test_runner.py proposed by sipa in PR #23995.

Remove the num_running variable as it can be implied by the length of the jobs list.

Remove the i variable as it can be implied by the length of the test_results list.

Instead of counting results to determine if finished, make the queue object itself 
responsible (by looking at running jobs and jobs left).

kevkevinpal commented:
I tried this with `--failfast` and it seemed to properly fail fast

tdb3 commented:
> I tried this with `--failfast` and it seemed to properly fail fast

Thank you for the test

kevkevinpal commented:
I tested with this diff e82b3421a8d5fb82fe20e670857abbbe220ae09e

and ran `./test/functional/test_runner.py --failfast`

davidgumberg commented on test/functional/test_runner.py:
```diff
@@ -613,14 +613,12 @@ def run_tests(*, test_list, src_dir, build_dir, tmpdir, jobs=1, enable_coverage=
     max_len_name = len(max(test_list, key=len))
     test_count = len(test_list)
```
Maybe we can remove `test_count`? See below.

tdb3 replied:
`test_count` being assigned outside of the loop enables reporting of the correct total number of total tests on the right side of the `/`.

davidgumberg commented on test/functional/test_runner.py:
```diff
@@ -613,14 +613,12 @@ def run_tests(*, test_list, src_dir, build_dir, tmpdir, jobs=1, enable_coverage=
             test_results.append(test_result)
-            i += 1
-            done_str = "{}/{} - {}{}{}".format(i, test_count, BOLD[1], test_result.name, BOLD[0])
+            done_str = "{}/{} - {}{}{}".format(len(test_results), test_count, BOLD[1], test_result.name, BOLD[0])
```
Nit: We could use an f-string here and replace `test_count` with `len(test_list)`

```suggestion         
            done_str = f"{len(test_results)}/{len(test_list)} - {BOLD[1]}{test_result.name}{BOLD[0]}"

tdb3 replied:
Thanks, updating to an f-string better aligns with the style guidelines (and that line is being changed anyway).
Leaving other format strings untouched for now (those could be updated in another refactor PR).

tdb3 committed:
test: simplify test_runner.py

Remove the num_running variable as it can be implied by the
length of the jobs list.

Remove the i variable as it can be implied by the length of the
test_results list.

Instead of counting results to determine if finished, make the
queue object itself responsible (by looking at running jobs and
jobs left).

Originally proposed by @sipa in PR #23995.

Co-authored-by: Pieter Wuille <pieter@wuille.net>

tdb3 commented:
Thanks for the review @daviddumberg!
Force pushed.  Responses inline.

marcofleon reviewed (approved):
Tested ACK 8f2d814407c280f863be0303b4c2c1efdc0b8f36. 

Reviewed the code changes and I think they make the program a bit more readable. I ran `test_runner.py` and there were no issues. Also tested with `--failfast` and it did indeed fail fast at the expected test.

davidgumberg commented:
reACK https://github.com/bitcoin/bitcoin/commit/8f2d814407c280f863be0303b4c2c1efdc0b8f36

mzumsande reviewed (commented):
Code Review ACK 8f2d814407c280f863be0303b4c2c1efdc0b8f36

tdb3 commented:
> Tested ACK 8f2d814407c280f863be0303b4c2c1efdc0b8f36. 
> 
> Reviewed the code changes and I think they make the program a bit more readable. I ran `test_runner.py` and there were no issues. Also tested with `--failfast` and it did indeed fail fast at the expected test.

Thank you for the review @marcofleon!

tdb3 commented:
> Code Review ACK 8f2d814407c280f863be0303b4c2c1efdc0b8f36

Thank you for the review @mzumsande

tdb3 commented:
rebased

mzumsande commented:
re-ACK 0831b54

>rebased

Why? This looked ready for merge with 3 ACKs. A rebase invalidates existing ACKs, so once these exist it should only be done if there is a conflict (Drahtbot will complain) or a specific reason (such as. a silent conflict). The CI will always rebase on master anyway for its runs (which could be restarted without rebasing if really necessary).

tdb3 commented:
> re-ACK 0831b54
> 
> >rebased
> 
> Why? This looked ready for merge with 3 ACKs. A rebase invalidates existing ACKs, so once these exist it should only be done if there is a conflict (Drahtbot will complain) or a specific reason (such as. a silent conflict). The CI will always rebase on master anyway for its runs (which could be restarted without rebasing if really necessary).

Thank you for the insight, didn't realize that. I'll avoid that moving forward. The rationale was to bring it to head while reinitiating CI (MacOS brew issue).

davidgumberg commented:
reACK https://github.com/bitcoin/bitcoin/commit/0831b54dfca1b9e728295fff500215da14589fc0
Verified that the commit is unchanged with range-diff, and re-ran `make check`: 

```
$ git range-diff a945f09...8f2d814 1105aa4..0831b54
1:  8f2d814407 = 1:  0831b54dfc test: simplify test_runner.py
```

marcofleon reviewed (commented):
re-ACK 0831b54dfca1b9e728295fff500215da14589fc0

maflcko reviewed (approved):
lgtm
